### ISSUE TITLE
Show romanized text to Western users in Japan when name:en is missing

### ISSUE THREAD

**Issue by vbertola** [organicmaps/organicmaps#5139]
As a Western user with an Italian-language Android phone, I often use the app to view Japan (I actually go there every now and then). Organic Maps will correctly show me foreigner-friendly names, i.e. it will show name:en in place of name (which of course is generally written in Japanese characters) at small scale, then it will show name:en + name[:ja] at large scale.

However, in the map there are of course many points (shops, hotels, roads, attractions - whatever) that do not have a name:en, but only a Japanese string in name. A good number of them, however, have name:ja-Latn, which is the transliteration of the Japanese characters into Latin ones. So, if no name:en is found but name:ja-Latn exists, the app should use and display the name:ja-Latn string in the exact way as it would do with a name:en string; this would really make a difference for Western users.

Example (the one that brought me here): https://www.openstreetmap.org/node/5808925153 This place is actually indicated as "Kazurabashi" in all Western language guides and pages, and has no name:en but has this string as name:ja-Latn; currently, Organic Maps only displays "かずら橋", which most Western users won't be able to read. If the app displayed "Kazurabashi\nかずら橋", then the POI would be recognized at first sight. For further understanding, compare with the nearby waterfall https://www.openstreetmap.org/node/756335384 , which has "Biwa-no-taki" in name:en instead of name:ja-Latn; technically the tagging is incorrect (name:en should rather be "Biwa waterfall") but in Organic Maps this gives better results.

By the way, while you are at it, you could also think at what to do with name:ja-Hira, which is the representation in hiragana (phonetic) Japanese characters of the name in kanji (ideograms). Even many Japanese cannot recognize and read all kanji, so having ja-Hira somewhere could be useful even to Japanese users of the app; and for Westerners, in case no name:en and no name:ja-Latn exist, ja-Hira would still be easier to decipher than a kanji string (so it could be used as a further fallback in Western languages).

Oh, and if you wanted to be super compatible you could also look for name:ja_rm (deprecated alternative for name:ja-Latn) and name:ja_kana (deprecated alternative for name:ja-Hira). But that would possibly be too much.

In any case, at least showing ja-Latn would be a good improvement. Thanks!

**Comment by biodranik** [comment]
Thanks for the details, implementing it would be great.

However, there is a technical issue: we have a limited [hard-coded number of supported OSM languages](https://github.com/organicmaps/organicmaps/blob/master/coding/string_utf8_multilang.cpp#L15), specifically, 64. There are a few free language slots left there, but they should be used very carefully. Supporting more than 64 languages will require a (likely significant) refactoring. To implement a high-quality worldwide editor we likely need to do it anyway someday...

At the moment, the following related langs are supported:
- `name:ja`
- `name:ja_kana`
- `name:ja_rm`
- `name:en`
- `name:int_name`
- `name:alt_name`

So maybe we should do like this:

1. The first step can be renaming outdated tags and map them at the generation stage to their new versions: `name:ja_rm` => `name:ja-Latn`, and `name:ja_kana` => `name:ja-Hira`. After doing that, we'll have all old and new tags in mwm maps files.
2. The next step can be to add additional logic to fallback to ja-Latn, and then to ja-Hira as an alternative, international translation.

**Comment by vbertola** [comment]
Taginfo gives me this count:

1 007 527 [name:ja](https://taginfo.openstreetmap.org/keys/name%3Aja)
174 748 [name:ja-Hira](https://taginfo.openstreetmap.org/keys/name%3Aja-Hira)
108 442 [name:ja_rm](https://taginfo.openstreetmap.org/keys/name%3Aja_rm)
73 244 [name:ja-Latn](https://taginfo.openstreetmap.org/keys/name%3Aja-Latn)
54 326 [name:ja_kana](https://taginfo.openstreetmap.org/keys/name%3Aja_kana)

So, dropping ja_kana for ja-Hira is a clear advantage, but ja-Latn is still halfway in replacing ja_rm. But if I get your step 1 correctly, you would automatically change all "name:ja_rm" keys in "name:ja-Latn" (same for ja_kana) when you extract data from OSM to format them for use in the app, so that you do not have to support both of them in the app. If so, this would solve the problem of the duplicate keys.

As for the transliteration setting in the app, I don't really know how it works: I just tried to turn it off and I did not see any immediate change in the Japanese map I was viewing. However, you can just keep it as it works now; what I'm saying is that you should just take ja-Latn and maybe ja-Hira as an alternate transliteration when name:en does not exist.

**Comment by biodranik** [comment]
Can OSM's `name:int_name` be used for this purpose instead of `name:en`? Consider other users who don't speak English, but can read the Latin alphabet.

Btw, what are your device's preferred languages?

**Comment by vbertola** [comment]
> Can OSM's `name:int_name` be used for this purpose instead of `name:en`? Consider other users who don't speak English, but can read the Latin alphabet.

I didn't actually know "int_name" existed - I don't think it's used much. In fact, the Japan-only taginfo instance ( https://taginfo.openstreetmap.jp/search?q=int_name ) says that there are only 2486 int_name tags in the whole of Japan, two orders of magnitude less than the other tags.

Anyway, I would fallback to name:en (and then name:ja-Latn) even for other Western / Latin script languages, unless a more specific tag (i.e. name:es for Spanish or name:fr for French) can be found. In fact, this is what happens now somehow; my app is in Italian, yet I see name:en (I think I see name:it first, but it's also a very rare tag in Japan). In Japan, all signs in Latin script use either the English place names or the romanization (ja-Latn).

> Btw, what are your device's preferred languages?

In Android's settings? I only have Italian.

**Comment by peterjc** [comment]
I'd also like this functionality.

Perhaps you can do it while transforming the OSM data into your own format (i.e. transliterate/romanization to generate an "English" placename when not otherwise provided), or perhaps it can be done in the application on display (less data storage, but more processing on the device - and more special case code too)?

**Comment by biodranik** [comment]
We use ICU transliteration already. I assume that Japanese should be supported there, maybe we need to fix our transliteration code: https://unicode-org.github.io/icu/userguide/transforms/general/#scriptlanguage

**Comment by biodranik** [comment]
A great discussion about Hiragana, Katakana and Kanji difference: https://www.quora.com/How-can-you-tell-if-a-word-is-written-in-hiragana-katakana-and-kanji-without-knowing-hiragana-katakana-or-kanji-katakana-hiragana-Japanese

**Comment by biodranik** [comment]
Kanji [can't be transliterated by ICU](https://sourceforge.net/p/icu/mailman/icu-support/thread/OF88F58C7A.C5B946F5-ON85257848.002514D4-85257848.0026A99E%40lotus.com/#msg27144335).

There are some apps that are able to somehow process Kanji, mentioned on [this forum](https://jdorama.com/viewtopic_4437.htm) that can be used. One of them is [MeCab](https://en.wikipedia.org/wiki/MeCab).

**Comment by biodranik** [comment]
I also see in the debugger that for Japan we have only 'ja' language code, not ja_rm and ja_kana, despite their presence in the data. @vng should we add ja_rm and ja_kana to Japanese regions mwm headers?

Also, converting hiragana is not done now due to the 64 max languages limit (and you see that Japanese already takes 3 of them!) to katakana. That is a question to our Japanese users:

Would it be acceptable to only use katakana, and convert all hiragana to it? Or it is _critically important_ to have the original hirahana name for an object, and also allow to edit it, together with katakana, kanji and romanji?

**Comment by vbertola** [comment]
> I also see in the debugger that for Japan we have only 'ja' language code, not ja_rm and ja_kana, despite their presence in the data. @vng should we add ja_rm and ja_kana to Japanese regions mwm headers?
> 
> Also, converting hiragana is not done now due to the 64 max languages limit (and you see that Japanese already takes 3 of them!) to katakana. That is a question to our Japanese users:
> 
> Would it be acceptable to only use katakana, and convert all hiragana to it? Or it is _critically important_ to have the original hirahana name for an object, and also allow to edit it, together with katakana, kanji and romanji?

I am not Japanese but I think you got the question wrong. There are only (just...) three ways of writing a Japanese name:

1. the "normal" way of writing, which can contain a mix of kanji, hiragana and katakana, and goes into name:ja ;

2. the "phonetic" way of writing (furigana), which helps you to read the "normal" string if you don't know the kanji, and is normally done in hiragana only but could contain some katakana as well (definitely not kanji), and goes into name:ja-Hira (previously name:ja_kana, which is now deprecated and should just be considered equivalent to name:ja-Hira when found);

3. the "romanized" way of writing (romaji), which is the equivalent of the phonetic writing but in Latin characters, and goes into name:ja-Latn (previously name:ja_rm, now deprecated).

See also https://wiki.openstreetmap.org/wiki/Multilingual_names#Japan if this is not clear yet.

Bottom line: the same thing will never have a kanji name *and* a hiragana name *and* a katakana name, but it will have a normal name and (maybe) a phonetic name. Romanization is an extra for Westerners.

So, at most you have three language tags for Japan (plus name:en, which may contain a fourth name, like the actual translation into English of the name). If you want, you can allow the user to enter/edit all four of them. For the display, you could somehow reduce this to two: name:ja for Japanese users and (name:en or name:ja-Latn) for international users. You could then add name:ja-Hira, if existing, to the popup if the user selects the POI, in both cases. For ja_kana and ja_rm, they are deprecated and you should just convert them into ja-Hira and ja-Latn when found.
